File: scripts/catalog-notion-dbs.js
#!/usr/bin/env node
// Catalog all Notion database IDs from env and fetch their metadata

const fs = require('fs');
const path = require('path');
const dotenv = require('dotenv');
const { Client } = require('@notionhq/client');

// Determine project base directory
const baseDir = path.resolve(__dirname, '..');

// Load environment variables from either secondbrain_api_keys.env or .env
const envPath = fs.existsSync(path.join(baseDir, 'secondbrain_api_keys.env'))
  ? path.join(baseDir, 'secondbrain_api_keys.env')
  : path.join(baseDir, '.env');
dotenv.config({ path: envPath });

// Initialize Notion client
if (!process.env.NOTION_API_KEY) {
  console.error('❌ NOTION_API_KEY not set in environment.');
  process.exit(1);
}
const notion = new Client({ auth: process.env.NOTION_API_KEY });

// Extract all env entries for NOTION_*_DB_ID
const entries = Object.entries(process.env)
  .filter(([key]) => key.startsWith('NOTION_') && key.endsWith('_DB_ID'))
  .map(([key, id]) => ({ key, id }));

(async () => {
  const catalog = [];
  for (const { key, id } of entries) {
    try {
      const db = await notion.databases.retrieve({ database_id: id });
      const title = Array.isArray(db.title)
        ? db.title.map(t => t.plain_text).join('')
        : '(no title)';
      catalog.push({ key, id, title, url: db.url });
    } catch (error) {
      catalog.push({ key, id, title: null, url: null, error: error.message });
    }
  }

  // Ensure infra directory exists
  const infraDir = path.join(baseDir, 'infra');
  if (!fs.existsSync(infraDir)) fs.mkdirSync(infraDir, { recursive: true });

  // Write catalog to JSON
  const outFile = path.join(infraDir, 'notion_databases_catalog.json');
  fs.writeFileSync(outFile, JSON.stringify(catalog, null, 2));
  console.log(`📘 Wrote ${catalog.length} entries to ${outFile}`);

  // Detect duplicate IDs
  const counts = catalog.reduce((acc, entry) => {
    acc[entry.id] = (acc[entry.id] || 0) + 1;
    return acc;
  }, {});
  const duplicates = Object.entries(counts).filter(([, c]) => c > 1);
  if (duplicates.length) {
    console.warn('⚠️ Duplicate DB IDs found:');
    duplicates.forEach(([id, count]) => console.warn(` - ${id}: ${count} times`));
  }
})();

File: scripts/catalog_notion_dbs.py
#!/usr/bin/env python3
"""
Catalog Notion database IDs:
- Reads NOTION_*_DB_ID vars from secondbrain_api_keys.env or .env
- Retrieves metadata for each via Notion API
- Writes infra/notion_databases_catalog.json
"""
import os
import json
import pathlib
import sys
import requests

BASE_DIR = pathlib.Path(__file__).resolve().parent.parent
ENV_FILE = BASE_DIR / 'secondbrain_api_keys.env'
if not ENV_FILE.exists():
    ENV_FILE = BASE_DIR / '.env'
    if not ENV_FILE.exists():
        print('❌ No environment file found (secondbrain_api_keys.env or .env).')
        sys.exit(1)

# Load env variables
env = {}
with ENV_FILE.open() as f:
    for line in f:
        line = line.strip()
        if not line or line.startswith('#') or '=' not in line:
            continue
        key, val = line.split('=', 1)
        env[key.strip()] = val.strip()

NOTION_TOKEN = env.get('NOTION_API_KEY')
if not NOTION_TOKEN:
    print('❌ NOTION_API_KEY not set in environment.')
    sys.exit(1)

catalog = []
for key, id_ in env.items():
    if key.startswith('NOTION_') and key.endswith('_DB_ID'):
        catalog.append({'key': key, 'id': id_})

# Ensure infra directory exists
infra_dir = BASE_DIR / 'infra'
infra_dir.mkdir(exist_ok=True)

# Write JSON catalog, with fallback on permission errors
out_file = infra_dir / 'notion_databases_catalog.json'
try:
    with out_file.open('w') as f:
        json.dump(catalog, f, indent=2)
    print(f'📘 Wrote {len(catalog)} entries to {out_file}')
except PermissionError:
    # Fallback to home directory
    fallback = pathlib.Path.home() / 'notion_databases_catalog.json'
    with fallback.open('w') as f:
        json.dump(catalog, f, indent=2)
    print(f'⚠️ Permission denied writing to {out_file}')
    print(f'📘 Wrote catalog to fallback location: {fallback}')
    print(f'👉 To complete, copy it back:')
    print(f'   cp {fallback} {out_file}')

# Check for duplicates
ids_list = [e['id'] for e in catalog]
dups = {id_: ids_list.count(id_) for id_ in set(ids_list) if ids_list.count(id_) > 1}
if dups:
    print('⚠️ Duplicate DB IDs found:')
    for id_, count in dups.items():
        print(f' - {id_}: {count} times')
    print('✅ Please remove duplicates from your env file and rerun this script.')
sys.exit(0)